In [2]:
import pandas as pd

In [3]:
students_df = pd.read_csv("StudentPerformanceFactors.csv")
students_df.head()

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             23          84                  Low                High   
1             19          64                  Low              Medium   
2             24          98               Medium              Medium   
3             29          89                  Low              Medium   
4             19          92               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                         No            7               73              Low   
1                         No            8               59              Low   
2                        Yes            7               91           Medium   
3                        Yes            8               98           Medium   
4                        Yes            6               65           Medium   

  Internet_Access  Tutoring_Sessions Family_Income Teacher_Quality  \
0             Yes                  0           Low          Medium   
1             Yes                  2        Medium          Medium   
2             Yes                  2        Medium          Medium   
3             Yes                  1        Medium          Medium   
4             Yes                  3        Medium            High   

  School_Type Peer_Influence  Physical_Activity Learning_Disabilities  \
0      Public       Positive                  3                    No   
1      Public       Negative                  4                    No   
2      Public        Neutral                  4                    No   
3      Public       Negative                  4                    No   
4      Public        Neutral                  4                    No   

  Parental_Education_Level Distance_from_Home  Gender  Exam_Score  
0              High School               Near    Male          67  
1                  College           Moderate  Female          61  
2             Postgraduate               Near    Male          74  
3              High School           Moderate    Male          71  
4                  College               Near  Female          70

In [4]:
students_df.shape 

(6607, 20)

StudentPerformanceFactors.csv file have 6607 rows and 20 columns.

# Introduction

Selecting specific values of a pandas **DataFrame** or **Series** </br>
to work on is an implicit step in almost any data operation you will run,</br>
So one of the first things you need to learn in working with data in python is</br>
How to go about selecting the data point relevant to you quickly and effectively.

## 1) Accessing Value

### Native accessors 

Native python objects provide good ways of indexing data.</br>
pandas carries all of these over, Which helps make it easy to start with. </br>

In python, We can access the property of an object by accessing it's attrinute.</br>
A `book` object, For example, might have a `title` property, which we can accesss by calling `book.title`

`Columns` in a pandas `DataFrame` work in much that same way. </br>
Hence to access the `Attendance` column/property of the `students_df` we can use following code:-

In [5]:
students_df.Attendance

0       84
1       64
2       98
3       89
4       92
        ..
6602    69
6603    76
6604    90
6605    86
6606    67
Name: Attendance, Length: 6607, dtype: int64

### Access Column using indexing Operator [] 

If we have a `python dictionary`,</br>
We can access it's value using the indexing operator [ ].</br>
We can do the same with `columns` in a `DataFrame` :

In [6]:
students_df['Attendance']

0       84
1       64
2       98
3       89
4       92
        ..
6602    69
6603    76
6604    90
6605    86
6606    67
Name: Attendance, Length: 6607, dtype: int64

#### Difference between students_df.Attendance vs students_df['Attendance']

These are two ways of selecting a specific `Series` out of a `DataFrame`.</br>
Neither of them is more or less syntactically valid than the other, But </br>
the `index operator []` does have the `advantage` that it can handle `column name` with </br>
`space or reserved characters` in them.

(If we had a column name `Attendance of Students` column, `students_df.Attendance of Students` wouldn't work.)


#### Doesn't a pandas Series look kind of like a fancy dictionary? 
It pretty much is, so it's no surprise that, to drill down to a single specific value, </br>
we need only use the indexing operator [] once more:

In [7]:
students_df['Attendance'][0]

np.int64(84)

- Student in `0 index` has 84% of attendance.

## 2) Indexing in Pandas

The `indexing operators ([])` and `attributes selection (students_df.Attendance)` are nice because </br>
they work just like they do in the rest of python ecosystem. As a novice, This makes them easy to pick up and use.</br>
However, `Pandas` has it's own accessors operators, `loc` and `iloc`. </br>
For more advanced operations, These are the ones you are supposed to be using.

### A. Index-based selection: 

Pandas indexing works in one of two paradigms. </br>
The first is `index-based selection` : selecting data based on its numerical position in the data. </br>
`iloc` follows this paradigm.

To select the first `row` of data in a `DataFrame`, we may use the following code:

In [8]:
students_df.iloc[0]

Hours_Studied                          23
Attendance                             84
Parental_Involvement                  Low
Access_to_Resources                  High
Extracurricular_Activities             No
Sleep_Hours                             7
Previous_Scores                        73
Motivation_Level                      Low
Internet_Access                       Yes
Tutoring_Sessions                       0
Family_Income                         Low
Teacher_Quality                    Medium
School_Type                        Public
Peer_Influence                   Positive
Physical_Activity                       3
Learning_Disabilities                  No
Parental_Education_Level      High School
Distance_from_Home                   Near
Gender                               Male
Exam_Score                             67
Name: 0, dtype: object

Both `loc` and `iloc` are `row-first`, `column-second`.</br>
This is the opposite of what we do in native Python, which is column-first, row-second.</br>

This means that it's marginally easier to retrieve rows, and marginally harder to get retrieve columns. </br>
To get a column with `iloc`, we can do the following:

In [9]:
students_df.iloc[:, 0]

0       23
1       19
2       24
3       29
4       19
        ..
6602    25
6603    23
6604    20
6605    10
6606    15
Name: Hours_Studied, Length: 6607, dtype: int64

On its own, the `:` operator, which also comes from native `Python`, means `"everything"`.</br> 
When combined with other selectors, however, it can be used to indicate a `range of values`. </br>

`For example`,</br>
To select the `Hours_Studied` column from just the **first, second, and third row**, we would do:

In [10]:
students_df.iloc[0:3, 0]

0    23
1    19
2    24
Name: Hours_Studied, dtype: int64

**Or, to select just the second and third entries, we would do:**

In [11]:
students_df.iloc[1:3,0]

1    19
2    24
Name: Hours_Studied, dtype: int64

**It's also possible to pass a list:**

In [12]:
students_df.iloc[[1,2,3], 0]

1    19
2    24
3    29
Name: Hours_Studied, dtype: int64

In [13]:
students_df.iloc[[1,2,3], 0:4] 

Hours_Studied  Attendance Parental_Involvement Access_to_Resources
1             19          64                  Low              Medium
2             24          98               Medium              Medium
3             29          89                  Low              Medium

Finally, it's worth knowing that negative numbers can be used in selection. </br>
This will start counting forwards from the end of the values. </br>
So for example here are the last five elements of the dataset.

In [14]:
students_df.iloc[-5:]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
6602             25          69                 High              Medium   
6603             23          76                 High              Medium   
6604             20          90               Medium                 Low   
6605             10          86                 High                High   
6606             15          67               Medium                 Low   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
6602                         No            7               76   
6603                         No            8               81   
6604                        Yes            6               65   
6605                        Yes            6               91   
6606                        Yes            9               94   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
6602           Medium             Yes                  1          High   
6603           Medium             Yes                  3           Low   
6604              Low             Yes                  3           Low   
6605             High             Yes                  2           Low   
6606           Medium             Yes                  0        Medium   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
6602          Medium      Public       Positive                  2   
6603            High      Public       Positive                  2   
6604          Medium      Public       Negative                  2   
6605          Medium     Private       Positive                  3   
6606          Medium      Public       Positive                  4   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
6602                    No              High School               Near   
6603                    No              High School               Near   
6604                    No             Postgraduate               Near   
6605                    No              High School                Far   
6606                    No             Postgraduate               Near   

      Gender  Exam_Score  
6602  Female          68  
6603  Female          69  
6604  Female          68  
6605  Female          68  
6606    Male          64

### B. Label Based Selection 

The second paradigm for attribute selection is the one followed by the `loc` operator: label-based selection. </br>
In this paradigm, it's the data index value, not its position, which matters.

For example, to get the first entry in `student_df` of `Hours_Studied` column, we would now do the following:

In [15]:
students_df.loc[0, 'Hours_Studied']

np.int64(23)

`iloc` is conceptually simpler than `loc` because it ignores the dataset's indices.</br>
When we use `iloc` we treat the dataset like a big matrix (a list of lists), one that we have to index into by position.</br>
`loc`, by contrast, uses the information in the indices to do its work. </br>
Since your dataset usually has meaningful indices, it's usually easier to do things using `loc` instead.

For example, here's one operation that's much easier using `loc`:

In [16]:
students_df.loc[:, ['Hours_Studied', 'Attendance', 'Access_to_Resources','Peer_Influence','Previous_Scores','Exam_Score']]

Hours_Studied  Attendance Access_to_Resources Peer_Influence  \
0                23          84                High       Positive   
1                19          64              Medium       Negative   
2                24          98              Medium        Neutral   
3                29          89              Medium       Negative   
4                19          92              Medium        Neutral   
...             ...         ...                 ...            ...   
6602             25          69              Medium       Positive   
6603             23          76              Medium       Positive   
6604             20          90                 Low       Negative   
6605             10          86                High       Positive   
6606             15          67                 Low       Positive   

      Previous_Scores  Exam_Score  
0                  73          67  
1                  59          61  
2                  91          74  
3                  98          71  
4                  65          70  
...               ...         ...  
6602               76          68  
6603               81          69  
6604               65          68  
6605               91          68  
6606               94          64  

[6607 rows x 6 columns]

#### Choosing between loc and iloc

When choosing or transitioning between `loc` and `iloc`, there is one "gotcha" worth keeping in mind, </br>
which is that the two methods use slightly different indexing schemes.</br>

`iloc` uses the Python stdlib (Standard library) indexing scheme, where the first element of the range is included and the last one excluded. So `0:10` will select entries 0,...,9. 

`loc`, meanwhile, indexes inclusively. So `0:10` will select entries 0,...,10.

Why the change? Remember that `loc` can index any stdlib type: strings, for example. </br>
If we have a DataFrame with index values Apples, ..., Potatoes, ..., and </br>
we want to select "all the alphabetical fruit choices between Apples and Potatoes", </br>
then it's a lot more convenient to index df.loc['Apples':'Potatoes'] than </br>
it is to index something like df.loc['Apples', 'Potatoet'] (t coming after s in the alphabet).

This is particularly confusing when the DataFrame index is a simple numerical list, e.g. 0,...,1000.</br>
In this case `df.iloc[0:1000]` will return 1000 entries, while `df.loc[0:1000]` return 1001 of them! </br>
To get 1000 elements using loc, you will need to go one lower and ask for df.loc[0:999].

Otherwise, the semantics of using `loc` are the same as those for `iloc`.

#### 1. .loc (Label-based indexing)
- loc is used to select data by label or boolean condition.
- It includes the start and end labels in slicing.
- It allows you to access rows and columns by their actual index or column names.

#### 2. .iloc (Integer-based indexing)
- iloc is used to select data by position (integer index).
- It uses integer positions starting from 0 (like Python lists).
- It excludes the end index in slicing (like Python).

## Manipulating The Index

`Label-based selection` derives its power from the labels in the index. Critically, </br>
the index we use is not immutable. We can manipulate the index in any way we see fit.

The set_index() method can be used to do the job. Here is what happens when we `set_index()` to the `Attendance` column/field:

In [17]:
students_df.set_index('Attendance')

Hours_Studied Parental_Involvement Access_to_Resources  \
Attendance                                                           
84                     23                  Low                High   
64                     19                  Low              Medium   
98                     24               Medium              Medium   
89                     29                  Low              Medium   
92                     19               Medium              Medium   
...                   ...                  ...                 ...   
69                     25                 High              Medium   
76                     23                 High              Medium   
90                     20               Medium                 Low   
86                     10                 High                High   
67                     15               Medium                 Low   

           Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
Attendance                                                            
84                                 No            7               73   
64                                 No            8               59   
98                                Yes            7               91   
89                                Yes            8               98   
92                                Yes            6               65   
...                               ...          ...              ...   
69                                 No            7               76   
76                                 No            8               81   
90                                Yes            6               65   
86                                Yes            6               91   
67                                Yes            9               94   

           Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
Attendance                                                                     
84                      Low             Yes                  0           Low   
64                      Low             Yes                  2        Medium   
98                   Medium             Yes                  2        Medium   
89                   Medium             Yes                  1        Medium   
92                   Medium             Yes                  3        Medium   
...                     ...             ...                ...           ...   
69                   Medium             Yes                  1          High   
76                   Medium             Yes                  3           Low   
90                      Low             Yes                  3           Low   
86                     High             Yes                  2           Low   
67                   Medium             Yes                  0        Medium   

           Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
Attendance                                                                 
84                  Medium      Public       Positive                  3   
64                  Medium      Public       Negative                  4   
98                  Medium      Public        Neutral                  4   
89                  Medium      Public       Negative                  4   
92                    High      Public        Neutral                  4   
...                    ...         ...            ...                ...   
69                  Medium      Public       Positive                  2   
76                    High      Public       Positive                  2   
90                  Medium      Public       Negative                  2   
86                  Medium     Private       Positive                  3   
67                  Medium      Public       Positive                  4   

           Learning_Disabilities Parental_Education_Level Distance_from_Home  \
Attendance                                                         

This is useful if you can come up with an index for the dataset which is better than the current one.

## 3) Selecting

### Conditional Based Selection / Filtering

So far we've been indexing various strides of data, using structural properties of the `DataFrame` itself. </br>
To do interesting things with the data, however, we often need to ask questions based on conditions.</br>

For example, suppose that we're interested specifically in `Exam_Score` of students who have Scored more than 70 and has studied more than or equals to 20 hours.</br>

We can start by checking if each `Exam_Score` of student and if `Hours_Studied` is greater eqauls to 20 or not:

In [18]:
students_df['Hours_Studied'] >= 20

0        True
1       False
2        True
3        True
4       False
        ...  
6602     True
6603     True
6604     True
6605    False
6606    False
Name: Hours_Studied, Length: 6607, dtype: bool

This operation produced a Series of True/False booleans based on the `Hours_Studied` of each record. </br> This result can then be used inside of `loc` to select the relevant data:

In [19]:
students_df.loc[students_df['Hours_Studied'] >= 20]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0                23          84                  Low                High   
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
6                29          84               Medium                 Low   
7                25          78                  Low                High   
...             ...         ...                  ...                 ...   
6599             30          94               Medium                 Low   
6601             20          83               Medium                 Low   
6602             25          69                 High              Medium   
6603             23          76                 High              Medium   
6604             20          90               Medium                 Low   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
0                            No            7               73   
2                           Yes            7               91   
3                           Yes            8               98   
6                           Yes            7               68   
7                           Yes            6               50   
...                         ...          ...              ...   
6599                         No            5               52   
6601                         No            6               51   
6602                         No            7               76   
6603                         No            8               81   
6604                        Yes            6               65   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
0                 Low             Yes                  0           Low   
2              Medium             Yes                  2        Medium   
3              Medium             Yes                  1        Medium   
6                 Low             Yes                  1           Low   
7              Medium             Yes                  1          High   
...               ...             ...                ...           ...   
6599              Low              No                  3          High   
6601              Low             Yes                  2        Medium   
6602           Medium             Yes                  1          High   
6603           Medium             Yes                  3           Low   
6604              Low             Yes                  3           Low   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
0             Medium      Public       Positive                  3   
2             Medium      Public        Neutral                  4   
3             Medium      Public       Negative                  4   
6             Medium     Private        Neutral                  2   
7               High      Public       Negative                  2   
...              ...         ...            ...                ...   
6599          Medium     Private        Neutral                  2   
6601          Medium      Public        Neutral                  4   
6602          Medium      Public       Positive                  2   
6603            High      Public       Positive                  2   
6604          Medium      Public       Negative                  2   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
0                       No              High School               Near   
2                       No             Postgraduate               Near   
3                       No              High School           Moderate   
6                       No              High School           Moderate   
7                       No              High School                Far   
...                    ...                      ...                ...   
6599                    No             Postgraduate           Mo

This DataFrame has ~3528 rows. The original had ~6607. That means that around 53% of students have studied 20 or more hours.

We also wanted to know the `Exam_Score` of students who have scored more than 70 and studied 20 or more hours.

We can use the `ampersand (&)` to bring the two questions together:

In [20]:
students_df.loc[(students_df['Hours_Studied'] >= 20) & (students_df['Exam_Score'] > 70)]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
9                23          98               Medium              Medium   
27               22          83                 High                High   
55               26          88               Medium                High   
...             ...         ...                  ...                 ...   
6565             24          89               Medium                 Low   
6566             29          96                 High              Medium   
6572             33          95                  Low                 Low   
6592             29         100               Medium                 Low   
6595             28          78               Medium                High   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
2                           Yes            7               91   
3                           Yes            8               98   
9                           Yes            8               71   
27                          Yes            6               94   
55                          Yes            5               79   
...                         ...          ...              ...   
6565                         No            4               99   
6566                         No            8               85   
6572                        Yes            6               77   
6592                        Yes            8              100   
6595                         No            7               92   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
2              Medium             Yes                  2        Medium   
3              Medium             Yes                  1        Medium   
9              Medium             Yes                  0          High   
27             Medium             Yes                  0          High   
55             Medium             Yes                  1           Low   
...               ...             ...                ...           ...   
6565             High             Yes                  1          High   
6566              Low             Yes                  3          High   
6572              Low             Yes                  2          High   
6592              Low             Yes                  0           Low   
6595              Low             Yes                  1        Medium   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
2             Medium      Public        Neutral                  4   
3             Medium      Public       Negative                  4   
9               High      Public       Positive                  5   
27            Medium      Public        Neutral                  2   
55              High     Private       Positive                  4   
...              ...         ...            ...                ...   
6565            High     Private       Positive                  3   
6566          Medium      Public        Neutral                  3   
6572            High     Private       Positive                  3   
6592          Medium     Private        Neutral                  5   
6595            High      Public       Positive                  3   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
2                       No             Postgraduate               Near   
3                       No              High School           Moderate   
9                       No              High School           Moderate   
27                      No                  College           Moderate   
55                      No                  College           Moderate   
...                    ...                      ...                ...   
6565                    No             Postgraduate             

This `DataFrame` has only **900 rows** where the rows of students who have studied more or eqaul to **20 hours** is **3528** and the original DataFrame had **6607** rows.

That means :

- `Only around (900 students) 25% of students where able to score more than 70 from the students who have studied more or eqaul to 20 hours.`


In [31]:
students_df.loc[students_df['Exam_Score']>70]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
5                19          88               Medium              Medium   
9                23          98               Medium              Medium   
11               17          97               Medium                High   
...             ...         ...                  ...                 ...   
6565             24          89               Medium                 Low   
6566             29          96                 High              Medium   
6572             33          95                  Low                 Low   
6592             29         100               Medium                 Low   
6595             28          78               Medium                High   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
2                           Yes            7               91   
3                           Yes            8               98   
5                           Yes            8               89   
9                           Yes            8               71   
11                          Yes            6               87   
...                         ...          ...              ...   
6565                         No            4               99   
6566                         No            8               85   
6572                        Yes            6               77   
6592                        Yes            8              100   
6595                         No            7               92   

     Motivation_Level Internet_Access  Tutoring_Sessions  ... Teacher_Quality  \
2              Medium             Yes                  2  ...          Medium   
3              Medium             Yes                  1  ...          Medium   
5              Medium             Yes                  3  ...          Medium   
9              Medium             Yes                  0  ...            High   
11                Low             Yes                  2  ...            High   
...               ...             ...                ...  ...             ...   
6565             High             Yes                  1  ...            High   
6566              Low             Yes                  3  ...          Medium   
6572              Low             Yes                  2  ...            High   
6592              Low             Yes                  0  ...          Medium   
6595              Low             Yes                  1  ...            High   

     School_Type Peer_Influence Physical_Activity  Learning_Disabilities  \
2         Public        Neutral                 4                     No   
3         Public       Negative                 4                     No   
5         Public       Positive                 3                     No   
9         Public       Positive                 5                     No   
11       Private        Neutral                 2                     No   
...          ...            ...               ...                    ...   
6565     Private       Positive                 3                     No   
6566      Public        Neutral                 3                     No   
6572     Private       Positive                 3                     No   
6592     Private        Neutral                 5                    Yes   
6595      Public       Positive                 3                     No   

     Parental_Education_Level Distance_from_Home  Gender Exam_Score  \
2                Postgraduate               Near    Male         74   
3                 High School           Moderate    Male         71   
5                Postgraduate               Near    Male         71   
9                 High School           Moderate    Male         72   
11                High School               Near    Male         71   


- `Only around (1083 students out of 6607) 16% of students where able to score more than 70 from the DataFrame regradless of how many hours they have studied.`

In [22]:
students_df.loc[(students_df['Hours_Studied'] < 20) & (students_df['Exam_Score'] > 70)]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
5                19          88               Medium              Medium   
11               17          97               Medium                High   
21               19          99               Medium                High   
57               18          94               Medium              Medium   
88               17          86               Medium                High   
...             ...         ...                  ...                 ...   
6411             19          92                 High              Medium   
6487             12          97                 High                 Low   
6499             18          97                 High                High   
6522             18          90                 High                High   
6545             18          98                 High              Medium   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
5                           Yes            8               89   
11                          Yes            6               87   
21                           No            6               84   
57                          Yes            6               87   
88                           No            5               97   
...                         ...          ...              ...   
6411                         No            6               70   
6487                        Yes            6               96   
6499                        Yes            6               76   
6522                        Yes            6               54   
6545                        Yes            6               62   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
5              Medium             Yes                  3        Medium   
11                Low             Yes                  2           Low   
21             Medium             Yes                  1        Medium   
57               High             Yes                  1        Medium   
88               High             Yes                  3          High   
...               ...             ...                ...           ...   
6411             High             Yes                  2        Medium   
6487           Medium             Yes                  2           Low   
6499           Medium             Yes                  0        Medium   
6522              Low             Yes                  1        Medium   
6545              Low             Yes                  3        Medium   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
5             Medium      Public       Positive                  3   
11              High     Private        Neutral                  2   
21              High      Public        Neutral                  3   
57            Medium      Public       Positive                  4   
88            Medium      Public        Neutral                  3   
...              ...         ...            ...                ...   
6411          Medium     Private       Negative                  3   
6487            High      Public       Positive                  2   
6499            High      Public        Neutral                  5   
6522            High      Public       Negative                  3   
6545            High      Public       Positive                  3   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
5                       No             Postgraduate               Near   
11                      No              High School               Near   
21                      No              High School               Near   
57                      No              High School               Near   
88                      No                  College               Near   
...                    ...                      ...                ...   
6411                    No                      NaN             


- `Only around (183 students out of 6607) 2% of students have scored more than 70 in Exam even after studying less than 20 hours.`

**Suppose we need to find any student who have scored more than 70 or have studied more than 20 regradless of his/her grade,
For this we use a pipe (|):**

In [23]:
students_df.loc[(students_df['Hours_Studied'] >= 20) | (students_df['Exam_Score'] > 70)]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0                23          84                  Low                High   
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
5                19          88               Medium              Medium   
6                29          84               Medium                 Low   
...             ...         ...                  ...                 ...   
6599             30          94               Medium                 Low   
6601             20          83               Medium                 Low   
6602             25          69                 High              Medium   
6603             23          76                 High              Medium   
6604             20          90               Medium                 Low   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
0                            No            7               73   
2                           Yes            7               91   
3                           Yes            8               98   
5                           Yes            8               89   
6                           Yes            7               68   
...                         ...          ...              ...   
6599                         No            5               52   
6601                         No            6               51   
6602                         No            7               76   
6603                         No            8               81   
6604                        Yes            6               65   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
0                 Low             Yes                  0           Low   
2              Medium             Yes                  2        Medium   
3              Medium             Yes                  1        Medium   
5              Medium             Yes                  3        Medium   
6                 Low             Yes                  1           Low   
...               ...             ...                ...           ...   
6599              Low              No                  3          High   
6601              Low             Yes                  2        Medium   
6602           Medium             Yes                  1          High   
6603           Medium             Yes                  3           Low   
6604              Low             Yes                  3           Low   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
0             Medium      Public       Positive                  3   
2             Medium      Public        Neutral                  4   
3             Medium      Public       Negative                  4   
5             Medium      Public       Positive                  3   
6             Medium     Private        Neutral                  2   
...              ...         ...            ...                ...   
6599          Medium     Private        Neutral                  2   
6601          Medium      Public        Neutral                  4   
6602          Medium      Public       Positive                  2   
6603            High      Public       Positive                  2   
6604          Medium      Public       Negative                  2   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
0                       No              High School               Near   
2                       No             Postgraduate               Near   
3                       No              High School           Moderate   
5                       No             Postgraduate               Near   
6                       No              High School           Moderate   
...                    ...                      ...                ...   
6599                    No             Postgraduate           Mo

- `Out of 6607 students, 3771 has studied more or equal to 20 hours or have scored more than 70 in Exam.`

- `Around 56% of students has studied more or equal to 20 hours or have scored more than 70 in Exam regardless of studying more or less than 20 hours.`

`Pandas` comes with a `few built-in conditional selectors`, two of which we will highlight here.

#### isin()
The first is `isin`. `isin` is lets you select data whose value "is in" a list of values. </br>
For example, here's how we can use it to select `Parental_Involvement`only with `Low` and `Medium`:

In [24]:
students_df.loc[students_df.Parental_Involvement.isin(['Low','Medium'])]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0                23          84                  Low                High   
1                19          64                  Low              Medium   
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
4                19          92               Medium              Medium   
...             ...         ...                  ...                 ...   
6599             30          94               Medium                 Low   
6600             12          98               Medium                 Low   
6601             20          83               Medium                 Low   
6604             20          90               Medium                 Low   
6606             15          67               Medium                 Low   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
0                            No            7               73   
1                            No            8               59   
2                           Yes            7               91   
3                           Yes            8               98   
4                           Yes            6               65   
...                         ...          ...              ...   
6599                         No            5               52   
6600                        Yes            4               54   
6601                         No            6               51   
6604                        Yes            6               65   
6606                        Yes            9               94   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
0                 Low             Yes                  0           Low   
1                 Low             Yes                  2        Medium   
2              Medium             Yes                  2        Medium   
3              Medium             Yes                  1        Medium   
4              Medium             Yes                  3        Medium   
...               ...             ...                ...           ...   
6599              Low              No                  3          High   
6600           Medium             Yes                  2        Medium   
6601              Low             Yes                  2        Medium   
6604              Low             Yes                  3           Low   
6606           Medium             Yes                  0        Medium   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
0             Medium      Public       Positive                  3   
1             Medium      Public       Negative                  4   
2             Medium      Public        Neutral                  4   
3             Medium      Public       Negative                  4   
4               High      Public        Neutral                  4   
...              ...         ...            ...                ...   
6599          Medium     Private        Neutral                  2   
6600            High     Private        Neutral                  3   
6601          Medium      Public        Neutral                  4   
6604          Medium      Public       Negative                  2   
6606          Medium      Public       Positive                  4   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
0                       No              High School               Near   
1                       No                  College           Moderate   
2                       No             Postgraduate               Near   
3                       No              High School           Moderate   
4                       No                  College               Near   
...                    ...                      ...                ...   
6599                    No             Postgraduate           Mo

#### isnull() and notnull()
The second is `isnull` (and its companion `notnull`). </br>
These methods let you highlight values which are (or are not) empty (NaN).</br>
For example, to filter out `Sleep_Hours` entries in the dataset, here's what we would do:

In [25]:
students_df.loc[students_df.Sleep_Hours.notnull()]

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0                23          84                  Low                High   
1                19          64                  Low              Medium   
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
4                19          92               Medium              Medium   
...             ...         ...                  ...                 ...   
6602             25          69                 High              Medium   
6603             23          76                 High              Medium   
6604             20          90               Medium                 Low   
6605             10          86                 High                High   
6606             15          67               Medium                 Low   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
0                            No            7               73   
1                            No            8               59   
2                           Yes            7               91   
3                           Yes            8               98   
4                           Yes            6               65   
...                         ...          ...              ...   
6602                         No            7               76   
6603                         No            8               81   
6604                        Yes            6               65   
6605                        Yes            6               91   
6606                        Yes            9               94   

     Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
0                 Low             Yes                  0           Low   
1                 Low             Yes                  2        Medium   
2              Medium             Yes                  2        Medium   
3              Medium             Yes                  1        Medium   
4              Medium             Yes                  3        Medium   
...               ...             ...                ...           ...   
6602           Medium             Yes                  1          High   
6603           Medium             Yes                  3           Low   
6604              Low             Yes                  3           Low   
6605             High             Yes                  2           Low   
6606           Medium             Yes                  0        Medium   

     Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
0             Medium      Public       Positive                  3   
1             Medium      Public       Negative                  4   
2             Medium      Public        Neutral                  4   
3             Medium      Public       Negative                  4   
4               High      Public        Neutral                  4   
...              ...         ...            ...                ...   
6602          Medium      Public       Positive                  2   
6603            High      Public       Positive                  2   
6604          Medium      Public       Negative                  2   
6605          Medium     Private       Positive                  3   
6606          Medium      Public       Positive                  4   

     Learning_Disabilities Parental_Education_Level Distance_from_Home  \
0                       No              High School               Near   
1                       No                  College           Moderate   
2                       No             Postgraduate               Near   
3                       No              High School           Moderate   
4                       No                  College               Near   
...                    ...                      ...                ...   
6602                    No              High School             

There is no null value or NaN value in Dataframe.

In [26]:
students_df.columns

Index(['Hours_Studied', 'Attendance', 'Parental_Involvement',
       'Access_to_Resources', 'Extracurricular_Activities', 'Sleep_Hours',
       'Previous_Scores', 'Motivation_Level', 'Internet_Access',
       'Tutoring_Sessions', 'Family_Income', 'Teacher_Quality', 'School_Type',
       'Peer_Influence', 'Physical_Activity', 'Learning_Disabilities',
       'Parental_Education_Level', 'Distance_from_Home', 'Gender',
       'Exam_Score'],
      dtype='object')

### Assigning Data

In [27]:
# Creating Assignment named column in DataFrame and assigning 'Yes' value in every entries.
students_df['Assignment'] = 'Yes'
students_df.head()

Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             23          84                  Low                High   
1             19          64                  Low              Medium   
2             24          98               Medium              Medium   
3             29          89                  Low              Medium   
4             19          92               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                         No            7               73              Low   
1                         No            8               59              Low   
2                        Yes            7               91           Medium   
3                        Yes            8               98           Medium   
4                        Yes            6               65           Medium   

  Internet_Access  Tutoring_Sessions  ... Teacher_Quality School_Type  \
0             Yes                  0  ...          Medium      Public   
1             Yes                  2  ...          Medium      Public   
2             Yes                  2  ...          Medium      Public   
3             Yes                  1  ...          Medium      Public   
4             Yes                  3  ...            High      Public   

  Peer_Influence Physical_Activity  Learning_Disabilities  \
0       Positive                 3                     No   
1       Negative                 4                     No   
2        Neutral                 4                     No   
3       Negative                 4                     No   
4        Neutral                 4                     No   

  Parental_Education_Level Distance_from_Home  Gender Exam_Score  Assignment  
0              High School               Near    Male         67         Yes  
1                  College           Moderate  Female         61         Yes  
2             Postgraduate               Near    Male         74         Yes  
3              High School           Moderate    Male         71         Yes  
4                  College               Near  Female         70         Yes  

[5 rows x 21 columns]

Create a DataFrame `top_oceania_wines` containing all reviews with at least 95 points (out of 100) for wines from Australia or New Zealand.

In [32]:
# Question From Kaggle

# Filter for students_df with more than 70 Exam_Score and from the students whose Parental_Education_Level is College or PostGraudate.

highScore_Student_with_Educated_parents = students_df[
    (students_df['Exam_Score'] > 70) & 
    (students_df['Parental_Education_Level'].isin(['College', 'Postgraduate']))
]

highScore_Student_with_Educated_parents


Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
2                24          98               Medium              Medium   
5                19          88               Medium              Medium   
27               22          83                 High                High   
55               26          88               Medium                High   
64               25          98                  Low                High   
...             ...         ...                  ...                 ...   
6516             30          78               Medium                High   
6549             23          90               Medium              Medium   
6565             24          89               Medium                 Low   
6566             29          96                 High              Medium   
6595             28          78               Medium                High   

     Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
2                           Yes            7               91   
5                           Yes            8               89   
27                          Yes            6               94   
55                          Yes            5               79   
64                          Yes            8               56   
...                         ...          ...              ...   
6516                        Yes            8               84   
6549                         No            7               89   
6565                         No            4               99   
6566                         No            8               85   
6595                         No            7               92   

     Motivation_Level Internet_Access  Tutoring_Sessions  ... Teacher_Quality  \
2              Medium             Yes                  2  ...          Medium   
5              Medium             Yes                  3  ...          Medium   
27             Medium             Yes                  0  ...          Medium   
55             Medium             Yes                  1  ...            High   
64             Medium              No                  0  ...            High   
...               ...             ...                ...  ...             ...   
6516           Medium             Yes                  1  ...             Low   
6549              Low             Yes                  2  ...          Medium   
6565             High             Yes                  1  ...            High   
6566              Low             Yes                  3  ...          Medium   
6595              Low             Yes                  1  ...            High   

     School_Type Peer_Influence Physical_Activity  Learning_Disabilities  \
2         Public        Neutral                 4                     No   
5         Public       Positive                 3                     No   
27        Public        Neutral                 2                     No   
55       Private       Positive                 4                     No   
64       Private        Neutral                 3                     No   
...          ...            ...               ...                    ...   
6516     Private       Negative                 1                     No   
6549      Public       Positive                 4                     No   
6565     Private       Positive                 3                     No   
6566      Public        Neutral                 3                     No   
6595      Public       Positive                 3                     No   

     Parental_Education_Level Distance_from_Home  Gender Exam_Score  \
2                Postgraduate               Near    Male         74   
5                Postgraduate               Near    Male         71   
27                    College           Moderate    Male         71   
55                    College           Moderate    Male         72   
64               Postgraduate               Near    Male         71   


- Out of 6607 students, Only 622 students (around 9%) has scored more than 70 in Exams and Has Parent with higher Education atleast college or postgraduate.
- Out of 1087, 622 students (around 57%) has Parents with higher education, Among the students who has scored more than 70 in exams.
